.circleci/config.yml

```diff
@@ -1,200 +1,259 @@
     parameters:
-      dimod-version:
-        type: string
       python-version:
         type: string
-    docker:
-      - image: circleci/python:<< parameters.python-version >>
+      pip-constraints:
+        type: string
 
-    working_directory: ~/repo
+    docker:
+      - image: cimg/python:<< parameters.python-version >>
 
     steps:
-
       - checkout
 
-      - restore_cache:
-          keys:
-          - v2-dependencies-{{ checksum "requirements.txt" }}-{{ checksum "tests/requirements.txt" }}-{{ checksum "docs/requirements.txt" }}-{{ .Environment.CIRCLE_JOB }}
+      - restore_cache: &restore-cache-env
+          key: v1-dependencies-{{ checksum "requirements.txt" }}-{{ checksum "tests/requirements.txt" }}-{{ .Environment.CIRCLE_JOB }}
 
-      - run:
-          name: create virtualenv
+      - run: &create-virtualenv
+          name: Create virtual environment
           command: |
-            python -m virtualenv env
+            python -m venv env
 
-      - run:
-          name: install dependencies
+      - run: &install-requirements
+          name: Install requirements
           command: |
             . env/bin/activate
-            python --version
-            pip install -r requirements.txt -r tests/requirements.txt -r docs/requirements.txt
-            pip install --upgrade 'dimod<< parameters.dimod-version >>'
+            pip install -U pip wheel twine setuptools
+            pip install -r requirements.txt -r tests/requirements.txt
 
-      - save_cache:
+      - save_cache: &save-cache-env
+          key: v1-dependencies-{{ checksum "requirements.txt" }}-{{ checksum "tests/requirements.txt" }}-{{ .Environment.CIRCLE_JOB }}
           paths:
-            - ./env
-          key: v2-dependencies-{{ checksum "requirements.txt" }}-{{ checksum "tests/requirements.txt" }}-{{ checksum "docs/requirements.txt" }}-{{ .Environment.CIRCLE_JOB }}
+            - ~/.cache/pip
 
-      - run: &run-tests-template
-          name: run unittests
-          command: |
-            . env/bin/activate
-            python --version
-            coverage run -m unittest discover
+      - run: &build-package
+          name: Build package
+          command: env/bin/python setup.py sdist bdist_wheel
 
       - run:
-          name: codecov
+          name: Install package (with constraints)
           command: |
             . env/bin/activate
-            codecov
+            pip install . -c <(printf '%s\n' << parameters.pip-constraints >>)
 
-      - run:
-          name: doctest
-          command: |
-            . env/bin/activate
-            make -C docs/ doctest
+      - run: &run-python-tests
+          name: Run Python tests
+          command: env/bin/coverage run -m unittest discover
 
       - run:
-          name: linkcheck
+          name: Upload code coverage
           command: |
             . env/bin/activate
-            make -C docs/ linkcheck
+            codecov     # calls `coverage xml`, so we must activate venv
```

Comment:
Maybe you want to use the new [codecov uploader](https://docs.codecov.com/docs/codecov-uploader), and their [orb](https://github.com/dwavesystems/dwave-cloud-client/blob/5829a538b2045c9fac1bec7fe7888feb6973945d/.circleci/config.yml#L75-L76). But it looks they backed down from the hard deprecation of old uploaders.